### PR TITLE
Feature: Added 'compress' attribute option to the properties window

### LINKED ISSUE #15976: Feature: Add 'compress' attribute to folder/file
### What feature or improvement do you think would benefit Files?

Tried to set extended attribute 'compressed' to big text file, and didn't find how to set this attribute. In properties of the file i can set only 'read only' and 'hidden'.

### Requirements

- Add checkbox to attributes section for "compress" (only for NTFS file systems)

### Files Version

3.6.0.0

### Windows Version

22635.4005

### Comments

_No response_

### PR DESCRIPTION
**Pull request by d2dyno1**
**Summary**
This PR continues on the work made in the [ya/CompressedAttribute](https://github.com/files-community/Files/tree/ya/CompressedAttribute) branch and fixes the broken logic to actually enable/disable compression for items

FYI @yair100 

**Resolved / Related Issues**
- Closes [#15976](https://github.com/files-community/Files/issues/15976)

**Screenshots**
image

**Todo**

- [x] Tested setting compression for files
- [x] Tested setting compression for folders
- [x] Add a GetVolumeInformation check to see if volume supports compression through `FILE_FILE_COMPRESSION` attribute

### CODE REVIEW

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs [hunk 1]
```diff
@@ -935,6 +937,40 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
 			return Win32PInvoke.SetFileAttributesFromApp(lpFileName, lpFileInfo.dwFileAttributes & ~dwAttrs);
 		}
 
+		public static unsafe bool SetCompressionAttributeIoctl(string lpFileName, bool isCompressed)
+		{
+			using var hFile = PInvoke.CreateFile(
+				lpFileName,
+				Win32PInvoke.GENERIC_READ | Win32PInvoke.GENERIC_WRITE,
```
Comment (0x5bfa) [inline src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:944]:
```suggestion
				(uint)FILE_ACCESS_RIGHTS.FILE_WRITE_ATTRIBUES,
```

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs [hunk 2]
```diff
@@ -935,6 +937,40 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
...
+				FILE_SHARE_MODE.FILE_SHARE_READ | FILE_SHARE_MODE.FILE_SHARE_WRITE,
+				lpSecurityAttributes: null,
+				FILE_CREATION_DISPOSITION.OPEN_EXISTING,
+				FILE_FLAGS_AND_ATTRIBUTES.FILE_ATTRIBUTE_NORMAL,
+				hTemplateFile: null);
+
+			if (hFile.IsInvalid)
+				return false;
+
+			var bytesReturned = 0u;
```
Comment (0x5bfa) [inline src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:954]:
```suggestion
			uint bytesReturned = 0u;
```

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs [hunk 3]
```diff
@@ -935,6 +937,40 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
...
+			var compressionFormat = isCompressed
+				? Win32PInvoke.COMPRESSION_FORMAT_DEFAULT
+				: Win32PInvoke.COMPRESSION_FORMAT_NONE;
```
Comment (0x5bfa) [inline src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:957]:
```suggestion
			COMPRESSION_FORMAT compressionFormat = isCompressed
				? COMPRESSION_FORMAT.COMPRESSION_FORMAT_DEFAULT
				: COMPRESSION_FORMAT.COMPRESSION_FORMAT_NONE;
```

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs [hunk 4]
```diff
@@ -935,6 +937,40 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
...
+			var result = PInvoke.DeviceIoControl(
+				new(hFile.DangerousGetHandle()),
+				Win32PInvoke.FSCTL_SET_COMPRESSION,
+				&compressionFormat,
+				sizeof(ushort),
+				(void*)IntPtr.Zero,
+				0u,
+				&bytesReturned);
+
+			if (result.Value == 0)
```
Comment (0x5bfa) [inline src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:968]: BOOL is a struct and has implicit operator overloads for conversion between BOOL and bool.
```suggestion
			if (!result)
```

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs [hunk 5]
```diff
@@ -935,6 +937,40 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
...
+				return false;
+
+			var bytesReturned = 0u;
+			var compressionFormat = isCompressed
+				? Win32PInvoke.COMPRESSION_FORMAT_DEFAULT
+				: Win32PInvoke.COMPRESSION_FORMAT_NONE;
+
+			var result = PInvoke.DeviceIoControl(
+				new(hFile.DangerousGetHandle()),
+				Win32PInvoke.FSCTL_SET_COMPRESSION,
```
Comment (0x5bfa) [inline src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:961]:
```suggestion
				(uint)FSCTL_SET_COMPRESSION,
```

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs [hunk 6]
```diff
@@ -13,6 +13,8 @@
 using System.Windows.Forms;
 using Vanara.PInvoke;
 using Windows.System;
+using Windows.Win32;
+using Windows.Win32.Storage.FileSystem;
```
Comment (0x5bfa) [inline src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:17]:
```suggestion
using Windows.Win32.Storage.FileSystem;
using Windows.Win32.System.Ioctl;
```

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs [hunk 7]
```diff
@@ -935,6 +937,40 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
...
+			var compressionFormat = isCompressed
+				? Win32PInvoke.COMPRESSION_FORMAT_DEFAULT
+				: Win32PInvoke.COMPRESSION_FORMAT_NONE;
+
+			var result = PInvoke.DeviceIoControl(
+				new(hFile.DangerousGetHandle()),
+				Win32PInvoke.FSCTL_SET_COMPRESSION,
+				&compressionFormat,
+				sizeof(ushort),
+				(void*)IntPtr.Zero,
```
Comment (0x5bfa) [inline src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:964]: I believe "null" works here.
```suggestion
				null,
```

File: src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs
Comment (0x5bfa) [inline src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs:1]: Please revert changes in this file.

File: src/Files.App/ViewModels/Properties/Items/CombinedProperties.cs [hunk 8]
```diff
@@ -170,6 +170,28 @@ private void ViewModel_PropertyChanged(object sender, System.ComponentModel.Prop
...
+								});
+							}
+							else
+							{
+								List.ForEach(x =>
+								{
+									Win32Helper.SetCompressionAttributeIoctl(x.ItemPath, false);
+								});
+							}
+						}
```
Comment (0x5bfa) [inline src/Files.App/ViewModels/Properties/Items/CombinedProperties.cs:192]:
```suggestion
						if (ViewModel.IsContentCompressed ?? false)
						{
							List.ForEach(x =>
							{
								Win32Helper.SetCompressionAttributeIoctl(x.ItemPath, true);
							});
						}
						else
						{
							List.ForEach(x =>
							{
								Win32Helper.SetCompressionAttributeIoctl(x.ItemPath, false);
							});
						}
```

File: src/Files.App/ViewModels/Properties/Items/FileProperties.cs [hunk 9]
```diff
@@ -279,13 +281,23 @@ private async void ViewModel_PropertyChanged(object sender, System.ComponentMode
...
 					break;
 
+				case nameof(ViewModel.IsContentCompressed):
+					if (ViewModel.IsContentCompressed is not null)
+					{
+						if ((bool)ViewModel.IsContentCompressed)
+							Win32Helper.SetCompressionAttributeIoctl(Item.ItemPath, true);
+						else
+							Win32Helper.SetCompressionAttributeIoctl(Item.ItemPath, false);
+					}
```
Comment (0x5bfa) [inline src/Files.App/ViewModels/Properties/Items/FileProperties.cs:298]:
```suggestion
					Win32Helper.SetCompressionAttributeIoctl(Item.ItemPath, ViewModel.IsContentCompressed ?? false);
```

File: src/Files.App/ViewModels/Properties/Items/FolderProperties.cs [hunk 10]
```diff
@@ -211,6 +214,16 @@ private async void ViewModel_PropertyChanged(object sender, System.ComponentMode
...
 					break;
 
+				case "IsContentCompressed":
+					if (ViewModel.IsContentCompressed is not null)
+					{
+						if ((bool)ViewModel.IsContentCompressed)
+							Win32Helper.SetCompressionAttributeIoctl(Item.ItemPath, true);
+						else
+							Win32Helper.SetCompressionAttributeIoctl(Item.ItemPath, false);
+					}
```
Comment (0x5bfa) [inline src/Files.App/ViewModels/Properties/Items/FolderProperties.cs:224]:
```suggestion
					Win32Helper.SetCompressionAttributeIoctl(Item.ItemPath, ViewModel.IsContentCompressed ?? false);
```

File: src/Files.App/ViewModels/Properties/Items/LibraryProperties.cs [hunk 11]
```diff
@@ -161,6 +162,16 @@ private void ViewModel_PropertyChanged(object sender, System.ComponentModel.Prop
...
 					break;
+
+				case "IsContentCompressed":
+					if (ViewModel.IsContentCompressed is not null)
+					{
+						if ((bool)ViewModel.IsContentCompressed)
+							Win32Helper.SetCompressionAttributeIoctl(Library.ItemPath, true);
+						else
+							Win32Helper.SetCompressionAttributeIoctl(Library.ItemPath, false);
+					}
```
Comment (0x5bfa) [inline src/Files.App/ViewModels/Properties/Items/LibraryProperties.cs:173]:
```suggestion
					Win32Helper.SetCompressionAttributeIoctl(Item.ItemPath, ViewModel.IsContentCompressed ?? false);
```

**Review by 0x5bfa — COMMENTED** [review]
**Summary**
- Please use the constants win32metadata defines.
- Simplified some if-else's.

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs (hunk 1 above)
Comment (0x5bfa) [reply src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:944]: If it doesn't work well, let me know. I think standard write access works as well.

**Comment by d2dyno1** [comment]
@0x5bfa I implemented the changes you requested. Unfortunately, some of them didn't compile and CsWin32 didn't feel like SG-ing enums for me so I reverted to the old values.

File: src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs
Comment (d2dyno1) [reply src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs:1]: No (see comment)

**Comment by 0x5bfa** [comment]
Can you try to add "COMPRESSION_FORMAT" and "FILE_ACCESS_RIGHTS" in NativeMethods.txt beforehand and use these enums?

**Comment by yaira2** [comment]
@d2dyno1 fyi

File: src/Files.App/Data/Models/SelectedItemsPropertiesViewModel.cs [hunk 12]
```diff
@@ -718,6 +718,31 @@ public bool? IsHiddenEditedValue
 			set => SetProperty(ref isHiddenEditedValue, value);
 		}
 
+		private bool? isContentCompressed;
+		public bool? IsContentCompressed
```
Comment (0x5bfa) [inline src/Files.App/Data/Models/SelectedItemsPropertiesViewModel.cs:722]: Could you add a comment like "This is only for NTFS" so others wouldn't be confused it with archive files.

File: src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs [hunk 13]
```diff
@@ -45,8 +40,18 @@ public static partial class Win32PInvoke
...
-		public const int MAXIMUM_REPARSE_DATA_BUFFER_SIZE = 16 * 1024;
+		// FSCTL
 		public const int FSCTL_GET_REPARSE_POINT = 0x000900A8;
+		public const int FSCTL_LOCK_VOLUME = 0x00090018;
+		public const int FSCTL_DISMOUNT_VOLUME = 0x00090020;
+		public const uint FSCTL_SET_COMPRESSION = 0x9C040u;
+
+		// Compression
+		public const ushort COMPRESSION_FORMAT_NONE = 0x0000;
+		public const ushort COMPRESSION_FORMAT_DEFAULT = 0x0001;
```
Comment (0x5bfa) [inline src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs:51]: You can use COMPRESSION_FORMAT enum.

File: src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs [hunk 14]
```diff
@@ -45,8 +40,18 @@ public static partial class Win32PInvoke
 		public const uint OPEN_ALWAYS = 4;
 		public const uint TRUNCATE_EXISTING = 5;
 
-		public const int MAXIMUM_REPARSE_DATA_BUFFER_SIZE = 16 * 1024;
+		// FSCTL
 		public const int FSCTL_GET_REPARSE_POINT = 0x000900A8;
+		public const int FSCTL_LOCK_VOLUME = 0x00090018;
+		public const int FSCTL_DISMOUNT_VOLUME = 0x00090020;
+		public const uint FSCTL_SET_COMPRESSION = 0x9C040u;
```
Comment (0x5bfa) [inline src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs:47]: I think you can add these const names to CsWin32 and use them from the `Windows.Win32.PInvoke` class like `PInvoke.FSCTL_SET_COMPRESSION`.

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs [hunk 15]
```diff
@@ -11,8 +11,16 @@
 using System.Runtime.InteropServices.ComTypes;
 using System.Text;
 using System.Windows.Forms;
+using Vanara;
 using Vanara.PInvoke;
 using Windows.System;
+using Windows.Win32;
+using Windows.Win32.Foundation;
+using Windows.Win32.Storage.FileSystem;
```
Comment (0x5bfa) [inline src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:19]: If you struggled with co-existence of CsWin32 and Vanara, I don't mind to create a new file `Win32Helper.CsWin32.cs` that only contains CsWin32 references so we can do replacement of Vanara easily. However, as it depends upon you, I'll leave it up to you.

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs [hunk 16]
```diff
@@ -935,10 +943,72 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
 			return Win32PInvoke.SetFileAttributesFromApp(lpFileName, lpFileInfo.dwFileAttributes & ~dwAttrs);
 		}
 
+		public static unsafe bool CanCompressContent(string path)
+		{
+			path = Path.GetPathRoot(path) ?? string.Empty;
+
+			// MAX_PATH = 260 + 1 (null termination)
```
Comment (0x5bfa) [inline src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:950]: `MAX_PATH` is 260 including null termination char (`\0`), and 3 (`C:\`) + 256 (= 259) can be used as path chars.

```c
// Definition
#define MAX_PATH 260

// Usage ("" of C/C++/C# appends \0 automatically and if you use all 260, this'd stack-overflow)
char szPath[MAX_PATH] = "C:\\Windows";
```

```suggestion
			// MAX_PATH = 260 (includes null termination char)
```

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs [hunk 17]
```diff
@@ -935,10 +943,72 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
...
+				&maxComponentLength,
+				&fileSystemFlags,
+				fileSystemName,
+				261u);
+
+			if (!success)
+				return false;
+
+			// 0x00000010 FILE_FILE_COMPRESSION
+			return (fileSystemFlags & 0x00000010) != 0;
```
Comment (0x5bfa) [inline src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:972]: You can add FILE_FILE_COMPRESSION to CsWin32 and use from `PInvoke.FILE_FILE_COMPRESSION`

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs [hunk 18]
```diff
@@ -935,10 +943,72 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
 			return Win32PInvoke.SetFileAttributesFromApp(lpFileName, lpFileInfo.dwFileAttributes & ~dwAttrs);
 		}
 
+		public static unsafe bool CanCompressContent(string path)
+		{
+			path = Path.GetPathRoot(path) ?? string.Empty;
+
+			// MAX_PATH = 260 + 1 (null termination)
+			var volumeName = new PWSTR();
+			var fileSystemName = new PWSTR();
```
Comment (0x5bfa) [inline src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:952]: Don't you have to allocate? If so, you can do as follows:

```c#
const int MAX_PATH = 260;
PWTSR pVolumeName = stackalloc char[MAX_PATH]; // you may need casting maybe?
PWSTR pFileSystemName = stackalloc char[MAX_PATH];
```

File: src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs (hunk 14 above)
Comment (d2dyno1) [reply src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs:47]: What's the enum name?

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs (hunk 16 above)
Comment (d2dyno1) [reply src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:950]: Good to know, thanks!

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs (hunk 17 above)
Comment (d2dyno1) [reply src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:972]: What's the enum name?

**Comment by d2dyno1** [comment]
I think we can move on with it @0x5bfa

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs (hunk 17 above)
Comment (0x5bfa) [reply src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:973]: This is not a member of an enum but a **constant** value in C.
You can add FILE_FILE_COMPRESSION and use it like PInvoke.FILE_FILE_COMPRESSION

File: src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs (hunk 14 above)
Comment (0x5bfa) [reply src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs:47]: As above.

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs (hunk 18 above)
Comment (0x5bfa) [reply src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:953]: In that cause, I think you can specify null to the dedicated parameter.

```c#
PInvoke.GetVolumeInformation(
    path,
    null,
    0u,
    null,
    null,
    &dwFileSystemFlags,
    null,
    0u);
```

File: src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs (hunk 13 above)
Comment (yaira2) [reply src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs]: This should be resolved with 31289160e0e62cb693cbe5d85fc6026355b58fc7

File: src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs (hunk 14 above)
Comment (yaira2) [reply src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs]: This should be resolved with 31289160e0e62cb693cbe5d85fc6026355b58fc7

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs (hunk 17 above)
Comment (yaira2) [reply src/Files.App/Helpers/Win32/Win32Helper.Storage.cs]: This should be resolved with 31289160e0e62cb693cbe5d85fc6026355b58fc7

File: src/Files.App/Helpers/Win32/Win32Helper.Storage.cs (hunk 18 above)
Comment (d2dyno1) [reply src/Files.App/Helpers/Win32/Win32Helper.Storage.cs]: Addressed in [`c50e313` (#16319)](https://github.com/files-community/Files/pull/16319/commits/c50e31335f13669d3619e4de79a20032a537160b)

**Comment by 0x5bfa** [comment]
Looks great.

**Review by yair100 — APPROVED** [review]

**Pull request merged**